fsspec/implementations/http.py

```diff
@@ -1,18 +1,54 @@
 from __future__ import print_function, division, absolute_import
 
+import aiohttp
+import asyncio
+import inspect
 import re
 import requests
+import weakref
 from urllib.parse import urlparse
 from fsspec import AbstractFileSystem
 from fsspec.spec import AbstractBufferedFile
 from fsspec.utils import tokenize, DEFAULT_BLOCK_SIZE
 from ..caching import AllBytes
+from ..utils import other_paths
 
 # https://stackoverflow.com/a/15926317/3821154
 ex = re.compile(r"""<a\s+(?:[^>]*?\s+)?href=(["'])(.*?)\1""")
 ex2 = re.compile(r"""(http[s]?://[-a-zA-Z0-9@:%_+.~#?&/=]+)""")
 
 
+def is_running_async():
+    for frame in inspect.stack(context=0):
+        if frame.function == "run_until_complete":
+            return True
+    return False
+
+
+def sync_wrapper(func):
```

Comment:
To make sure I understand the general idea: the sync version always calls the async implementation? So if I do something like

```python
fs = S3FSFileSystem()
fs.ls("test-bucket/")
```

That will eventually run code on the event loop?

Reply:
The only guarantee is that the original API is sync, i.e., returns concrete results; but it *may* be implemented on the event loop. Some backends (HDFS) won't work well with async, and it will take some time to add async to all those backends that will. 
You could instead say that it always runs something on the event loop, but some functions may block for a while, but I had not imagined it that way.

Reply:
What will happen if I'm doing some aysncio stuff on the event loop, and then I call some fsspec stuff from the main thread. Is it possible that some blocking IO ends up happening on the asyncio event loop, which will block the unrelated asyncio stuff I was doing?

Reply:
I *believe* that if you call something in fsspec, then coroutines will still process on the event loop while the caller waits for result; things get more complicated if any of the fsspec coroutines themselves call sync (e.g., via a method defined only on the superclass, which right now has nothing async in it).
Perhaps this could be tested by `cat`ing a large download while a periodic coroutine appnds timestamps to a list?

Reply:
(but you are getting to the root of my async knowledge, exactly why I want other people to look at this code!)

Reply:
I recommend stealing the approach taken by dask.distributed.  It handles this and a variety of other situations smoothly.

Reply:
You mean https://github.com/dask/distributed/blob/master/distributed/utils.py#L292 ? I don't know what's going on there, but it's essentially what I had in mind. However, shouldn't running in a separate thread be unnecessary? The given loop would, by default, be in the current thread.

Reply:
Just as a note, NJS recently pointed me to https://github.com/oremanj/greenback/ in a [thread](https://trio.discourse.group/t/concurrency-and-trio-as-implmentation-detail-in-blocking-sync-method/302/2) about reentrancy. It might be useful. 

Be careful if your main eventloop is running and the method you implement as "sync" are actually letting the event loop running you may actually break some assumptions.


When needing to "nest" eventloop, if it is a possibility, then you can also use a different runner (for example trio from within asyncio). That will really make your trio-syncified-method really sync (and blocking the outer eventloop) without the need to use threads.

Reply:
The Dask solution was designed to support simultaneous use in both async and sync situations, which I think was the situation that @TomAugspurger was getting at.